code/modules/mob/living/silicon/ai/ai_programs.dm

```diff
@@ -0,0 +1,1017 @@
+		var/damage_healed = 20 + (min(30, (10 * spell_level)))
+		target.heal_overall_damage(damage_healed, damage_healed)
+		if(spell_level >= 5)
+			for(var/obj/item/organ/external/E in target.bodyparts)
+				if(prob(5 * spell_level))
+					E.mend_fracture()
+					E.fix_internal_bleeding()
+					E.fix_burn_wound()
+
+/datum/spell/ai_spell/ranged/nanosurgeon_deployment/on_purchase_upgrade()
+	cooldown_handler.recharge_duration = max(min(base_cooldown, base_cooldown - (max(spell_level - 3, 0) * 60)), cooldown_min)
+
+// Enhanced Door Controls: Reduces delay in bolting and shocking doors
+/datum/ai_program/enhanced_doors
+	program_name = "Enhanced Door Controls"
+	program_id = "enhanced_doors"
+	description = "You enhance the subroutines that let you control doors, speeding up response times!"
+	nanite_cost = 0
+	unlock_text = "Doors connected and optimized. You feel right at home."
+	max_level = 5
+	upgrade = TRUE
+	/// Track the original delay
+	var/original_door_delay = 3 SECONDS
+
+/datum/ai_program/enhanced_doors/upgrade(mob/living/silicon/ai/user, first_install = FALSE)
+	..()
+	if(!istype(user))
+		return
+	user.door_bolt_delay = original_door_delay * (1 - (upgrade_level * 0.1))
+	installed = TRUE
+
+/datum/ai_program/enhanced_doors/downgrade(mob/living/silicon/ai/user)
+	..()
+	user.door_bolt_delay = original_door_delay * (1 - (upgrade_level * 0.1))
+
+/datum/ai_program/enhanced_doors/uninstall(mob/living/silicon/ai/user)
+	..()
+	user.door_bolt_delay = original_door_delay
+
+
+// Experimental Research Subsystem - Knowledge is power
+/datum/ai_program/research_subsystem
+	program_name = "Experimental Research Subsystem"
+	program_id = "research_subsystem"
+	description = "Put your processors to work spinning centrifuges and studying results. You unlock a new point of research in a random field."
+	cost = 5
+	nanite_cost = 60
+	power_type = /datum/spell/ai_spell/research_subsystem
+	unlock_text = "Research and Discovery submodule installation complete."
+
+/datum/spell/ai_spell/research_subsystem
+	name = "Experimental Research Subsystem"
+	desc = "Heal a crew member with large numbers of robotic nanomachines!"
+	action_icon = 'icons/obj/machines/research.dmi'
+	action_icon_state = "tdoppler"
+	auto_use_uses = FALSE
+	base_cooldown = 900 SECONDS
+	cooldown_min = 600 SECONDS
+	starts_charged = FALSE
+	level_max = 10
+	selection_activated_message = "<span class='notice'>You spool up your research tools...</span>"
+	var/rnd_server = "station_rnd"
+
+/datum/spell/ai_spell/research_subsystem/cast(list/targets, mob/user)
+	// First, find the RND server
+	var/network_manager_uid = null
+	for(var/obj/machinery/computer/rnd_network_controller/RNC in GLOB.rnd_network_managers)
+		if(RNC.network_name == rnd_server)
+			network_manager_uid = RNC.UID()
+			break
+	var/obj/machinery/computer/rnd_network_controller/RNC = locateUID(network_manager_uid)
+	if(!RNC) // Could not find the RND server. It probably blew up.
+		return
```

Comment:
the failure state here and below (`!files`) should probably return some message to the player letting them know why the spell failed